virtio-devices/src/vhost_user/net.rs

```diff
@@ -100,63 +104,58 @@ impl Net {
         let backend_features = vhost_user_net
             .get_features()
             .map_err(Error::VhostUserGetFeatures)?;
-        avail_features &= backend_features;
+        let acked_features = avail_features & backend_features;
         // Set features back is required by the vhost crate mechanism, since the
         // later vhost call will check if features is filled in master before execution.
         vhost_user_net
-            .set_features(avail_features)
+            .set_features(acked_features)
             .map_err(Error::VhostUserSetFeatures)?;
 
-        let mut protocol_features;
-        let mut acked_features = 0;
-        if avail_features & VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits() != 0 {
-            acked_features |= VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits();
-            protocol_features = vhost_user_net
-                .get_protocol_features()
-                .map_err(Error::VhostUserGetProtocolFeatures)?;
-        } else {
-            return Err(Error::VhostUserProtocolNotSupport);
-        }
-
-        protocol_features &=
+        let avail_protocol_features =
             VhostUserProtocolFeatures::MQ | VhostUserProtocolFeatures::CONFIGURE_MEM_SLOTS;
+        let backend_protocol_features =
+            if acked_features & VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits() != 0 {
+                vhost_user_net
+                    .get_protocol_features()
+                    .map_err(Error::VhostUserGetProtocolFeatures)?
+            } else {
+                return Err(Error::VhostUserProtocolNotSupport);
+            };
+        let acked_protocol_features = avail_protocol_features & backend_protocol_features;
 
         vhost_user_net
-            .set_protocol_features(protocol_features)
+            .set_protocol_features(acked_protocol_features)
             .map_err(Error::VhostUserSetProtocolFeatures)?;
 
-        let acked_protocol_features = protocol_features.bits();
+        // If the control queue feature has not been negotiated, let's decrease
+        // the number of queues and threads.
+        if acked_features & (1 << virtio_net::VIRTIO_NET_F_CTRL_VQ) == 0 {
+            num_queues -= 1;
+            num_threads -= 1;
+        }
```

Comment:
Well it's kinda hard to make it readable and simpler, since we have one thread per queue pair + one thread for the control queue.

Reply:
I realized we didn't need to know about the number of threads since there's no need for the `paused_sync` to be a barrier of num_threads + 1, so problem solved :)